### PR TITLE
Batches API

### PR DESCRIPTION
**Pull request by shakkhorpstu**

### CODE REVIEW

File: app/Helper/Constant.php [hunk 1]
```diff
@@ -0,0 +1,10 @@
+<?php
+
```
Comment (ManiruzzamanAkash) [inline app/Helper/Constant.php]: Use 
```php
declare(strict_types=1);
```

File: app/Http/Requests/BatchRequest.php [hunk 2]
```diff
@@ -0,0 +1,64 @@
+<?php
+
+namespace App\Http\Requests;
+
+use Illuminate\Foundation\Http\FormRequest;
```
Comment (ManiruzzamanAkash) [inline app/Http/Requests/BatchRequest.php]: Please remove this importing `FormRequest` class.
We're extending our own FormRequest in this namespace.

File: app/Repositories/BatchRepository.php [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+class BatchRepository extends CrudRepository
+{
+    protected string $model = Batch::class;
+
+    public function deleteBatch($batch): Batch
+    {
+        $batch->update(['deleted' => Constant::SOFT_DELETE]);
+        return $batch;
+    }
```
Comment (ManiruzzamanAkash) [inline app/Repositories/BatchRepository.php]: I would prefar move this to `CrudRepository` and name somthing like - `softDeleteItem()`

and maybe we'll there create other method for `hardDeleteItem()`


And then remove it from here.

File: app/Services/BatchService.php [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+        return $this->batchRepository->all();
+    }
+
+    public function getPaginatedBatches(?int $perPage = null, $filters = []): array
+    {
+        $filters['deleted'] = Constant::NOT_DELETED;
+        return $this->batchRepository->paginate($perPage, $filters);
+    }
+
+    public function getBatch(int $id)
```
Comment (ManiruzzamanAkash) [inline app/Services/BatchService.php]: Add return type

File: database/migrations/2024_04_01_074730_create_batches_table.php [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+     */
+    public function up(): void
+    {
+        Schema::create('batches', function (Blueprint $table) {
+            $table->id();
+            $table->unsignedBigInteger('faculty_id');
+            $table->string('name');
+            $table->string('title')->nullable();
+            $table->string('session', 20);
+            $table->unsignedInteger('total_student');
```
Comment (ManiruzzamanAkash) [inline database/migrations/2024_04_01_074730_create_batches_table.php]: default should be `0`

File: app/Helper/Constant.php [hunk 6]
```diff
@@ -0,0 +1,10 @@
...
+namespace App\Helper;
+
+class Constant
+{
+    CONST NOT_DELETED = 0;
+    CONST SOFT_DELETE = 1;
+    CONST HARD_DELETE = 9;
+}
```
Comment (ManiruzzamanAkash) [inline app/Helper/Constant.php]: I think this is very old approach before PHP introduces enum class.

https://enversanli.medium.com/how-to-use-enums-with-laravel-9-d18f1ee35b56

What I would suggest, create an Enums folder and add file `DeleteStatus.php`. and move these to there.

```php
enum DeleteStatus: int
{
    case NOT_DELETED = 0;
    case SOFT_DELETE = 1;
    case HARD_DELETE = 9;
}
```


And then use like - 
```php
DeleteStatus::SOFT_DELETE
```

File: app/Models/Batch.php [hunk 7]
```diff
@@ -0,0 +1,19 @@
+<?php
+
```
Comment (ManiruzzamanAkash) [inline app/Models/Batch.php:2]: Add 
```php
declare(strict_types=1);
```

File: app/Repositories/BatchRepository.php [hunk 8]
```diff
@@ -0,0 +1,17 @@
+<?php
+
```
Comment (ManiruzzamanAkash) [inline app/Repositories/BatchRepository.php:2]: Add 
```
declare(strict_types=1);
```

**Comment by ManiruzzamanAkash** [comment]
@shakkhorpstu Looks Good. Just couple of feedbacks

**Comment by ManiruzzamanAkash** [comment]
@shakkhorpstu 
@engr-shishir Added yesterday `CrudService`, maybe we would extend that while using Service. Then there would be less code.

File: app/Helper/Constant.php (hunk 1 above)
Comment (shakkhorpstu) [reply app/Helper/Constant.php]: solved

File: app/Http/Requests/BatchRequest.php (hunk 2 above)
Comment (shakkhorpstu) [reply app/Http/Requests/BatchRequest.php]: solved

File: app/Repositories/BatchRepository.php (hunk 3 above)
Comment (shakkhorpstu) [reply app/Repositories/BatchRepository.php]: solved

File: database/migrations/2024_04_01_074730_create_batches_table.php (hunk 5 above)
Comment (shakkhorpstu) [reply database/migrations/2024_04_01_074730_create_batches_table.php]: solved

File: app/Helper/Constant.php (hunk 6 above)
Comment (shakkhorpstu) [reply app/Helper/Constant.php]: solved

File: app/Repositories/BatchRepository.php (hunk 8 above)
Comment (shakkhorpstu) [reply app/Repositories/BatchRepository.php:2]: solved

File: app/Services/BatchService.php (hunk 4 above)
Comment (shakkhorpstu) [reply app/Services/BatchService.php]: solved

File: app/Models/Batch.php (hunk 7 above)
Comment (shakkhorpstu) [reply app/Models/Batch.php:2]: solved

**Pull request merged**